keps/sig-network/2091-cluster-network-policy/README.md

```diff
@@ -0,0 +1,1161 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Introduce new set of APIs to express an administrator's intent in securing
+their K8s cluster. This doc proposes the ClusterNetworkPolicy API to complement
+the developer focused NetworkPolicy API in Kubernetes.
+
+(TODO) In the multi-cluster context, we are thinking to name the extended NetworkPolicy
+to MultiClusterNetworkPolicy. To avoid the extension of this API resource to be named
+MultiCluster-ClusterNetworkPolicy, maybe we should name this resource differently
+to begin with? AdminNetworkPolicy is a popular alternative.
+
+## Motivation
+
+Kubernetes provides NetworkPolicy resources to control traffic within a
+cluster. NetworkPolicies focus on expressing a developers intent to secure
+their applications. Thus, in order to satisfy the needs of a security admin,
+we propose to introduce a new API that captures the administrators intent.
+
+### Goals
+
+The goals for this KEP are to satisfy the following key user stories:
+1. As a cluster administrator, I want to enforce irrevocable guardrails that
+   all workloads must adhere to in order to guarantee the safety of my clusters.
+   In particular I want to enforce certain network level access controls that are
+cluster scoped and cannot be overridden or bypassed by namespace scoped network policies.
+2. As a cluster administrator, I want to have the option to enforce network level access controls
+   that facilitate network multi-tenancy and strict network level isolation between multiple
+   teams and tenants sharing a cluster via use of namespaces or groupings of namespaces per tenant.
+3. As a cluster administrator, I want to optionally also deploy an additional default set of policies to all
+   workloads that may be overridden by the developers if needed.
+
+There are several unique properties that we need to add in order accomplish the
+user stories above.
+1. Deny rules and, therefore, hierarchical enforcement of policy
+2. Semantics for a cluster-scoped policy object that may include
+   namespaces/workloads that have not been created yet.
+3. Backwards compatibility with existing Kubernetes Network Policy API
+
+### Non-Goals
+
+Our mission is to solve the most common use cases that cluster admins have.
+That is, we don't want to solve for every possible policy permutation a user
+can think of.  Instead, we want to design an API that addresses 90-95% use cases
+while keeping the mental model easy to understand and use.
+The focus of this KEP is on cluster scoped controls for east-west traffic within
+a cluster. Cluster scoped controls for North-South traffic may be addressed via
+future versions of the api resources introduced in this or other future KEPs.
```

Comment:
I don't know if it goes here or not, but somewhere it should be mentioned explicitly that CNPs never affect north-south traffic, and thus also don't override or bypass NetworkPolicies with ipBlock rules that select external traffic

Reply:
Will add this

Reply:
It doesn't need to explicitly ignore it; CNP rules can only match traffic which has a pod as both its source and its destination, so any N-S traffic will necessarily not match any CNPs and thus not be acted upon.